### DISCUSSION TITLE
[VarExporter] PHP8.2 Deprecated dynamic properties created in LazyGhostTrait::__set()

### DISCUSSION THREAD

**Discussion by emteknetnz** [symfony/symfony#48638] [Q&A] [answered]
### Symfony version(s) affected

6.2.x

### Description

var-exporter [LazyGhostTrait::__set()](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/VarExporter/LazyGhostTrait.php#L256) contains `$this->$name = $value;` when `$scope` is `null` - use of [dynamic properties is deprecated](https://www.php.net/releases/8.2/en.php#deprecate_dynamic_properties) in PHP 8.2

This will probably also affect other methods in LazyGhostTrait that use dynamic properties such as `__get()`, `__isset()`, `__unset()`

### How to reproduce

```php
<?php

use Symfony\Component\VarExporter\LazyGhostTrait;

class MyObj
{
    use LazyGhostTrait;
}
```
You can replicate the deprecation warning with the following code in PHP 8.2:

```php
$obj = new MyObj();
$obj->abc = 123;
```

```
[Deprecated] Creation of dynamic property MyObj::$abc is deprecated
GET /

Line 236 in /var/www/vendor/symfony/var-exporter/LazyGhostTrait.php
```

### Possible Solution

Adding `#[\AllowDynamicProperties]` to LazyGhostTrait will NOT work - instead it will trigger the warning 
`[Alert] Cannot apply #[AllowDynamicProperties] to trait`

**Comment by nicolas-grekas** [comment]
The deprecation is legit: you cannot add a dynamic property to MyObj.

**Comment by emteknetnz** [comment]
It is a legit deprecation notice, however the notice is being triggered from `Line 236 in /var/www/vendor/symfony/var-exporter/LazyGhostTrait.php` inside `__set()` - this to me says that some of the logic in `LazyGhostTrait` is no longer viable?

To be fair, I really don't understand how `LazyGhostTrait` works or what it's even used for.

However `LazyGhostTrait` wants to add dynamic properties directly to an object via `$this->$name = $value;` on line 236.  `__set()` is still supported in PHP 8.2 so you could normally do something like

```php
private array $data = [];

public function __set($name, $value)
{
    // ...
    $this->data[$name] = $value;
}
```

I did attempt to create a PR along these lines, however `LazyGhostTrait` is very particular and I got endless unit test failures.

**Comment by nicolas-grekas** [comment, marked as answer]
The trait simulates what native PHP would do. Because you access the dynamic property, __set sets the property, and triggers the notice. This is the expected simulated behavior.

Then why are you trying to use it? What goal do you have?

LazyGhostTrait provides lazy-loading to a class. If you want this class to also have magic setters, the easiest way is to use inheritance, where the class that uses the trait extends the one with the magic methods.

**Comment by emteknetnz** [comment]
> Then why are you trying to use it? What goal do you have?

Goal was simply to get an open source project that uses symfony to be PHP 8.2 compatible.  I did a code scan looking for any dependencies that were using dynamic proprieties and I came across this.  I'm not sure if there's any code path within our own project that would actually trigger this deprecation warning.

Since you're confident that things are functioning as they should be there's probably nothing further that needs to be done here

Thank you for your prompt and helpful replies

**Comment by realjjaveweb** [comment]
If you found this trying to resolve this deprecated notice:
> Since symfony/var-exporter 6.4: Per-property lazy-initializers are deprecated and won't be supported anymore in 7.0, use an object initializer instead.

And you are using Doctrine, the solution is to update the package `doctrine/doctrine-bundle` to at least 2.11(.1) https://github.com/doctrine/DoctrineBundle/releases

**Reply by kcaporaso** [reply]
This may also have you upgrade your "doctrine/dbal" to at least 3.7.*
